### PR TITLE
Refactor strategy to handle `RecursionError` between `get_serializer_class`, `get_read_serializer`, and `get_write_serializer`

### LINKED ISSUE #86: Assertion error when used only with `get_serializer_class()` implementation
* drf-rw-serializers version: 1.2.0 + (I guess)
* Django version: 4.2.13
* Python version: 3.11.9
* Operating System: Linux / MacOS

### Description

We are currently in the middle of migration to RW serializers, meaning some viewsets has `get_read/write_serializer_class` implemented and some using old approach (1 serializer) which end-up being served from `get_serializer_class`. With the latest changes if the class implements `get_serializer_class` and does not implement `serializer_class` attribute, then assertion error raised.

### What I Did

Below I will try to post the simple code snippet
```python
class MyApiViewSet(drf_rw_serializers_viewsets.GenericViewSet[MyModel]):
    queryset = MyModel.objects.all()

    @typing_extensions.override
    def get_serializer_class(self) -> type[drf_serializers.BaseSerializer[MyModel]:
        serializer_class:type[drf_serializers.BaseSerializer[MyModel] | None = None
        if self.action == 'action_one':
            serializer_class = SomeSerializer
        if serializer_class is None:
            msg = f'Unknown action: {self.action}'
            raise ValueError(msg)
        return serializer_class
```

This snippet is very good to find out if you are missing any serializer for any action (especially it is done during schema generation).

This code was not producing any errors in `1.1.x` version but produces this in `1.2.x+`:
```
AssertionError: 'MyApiViewSet' should either include one of `serializer_class` and
`read_serializer_class` attribute, or override one of the `get_serializer_class()`,
`get_read_serializer_class()` method.
```

We do override `get_serializer_class` but apparently in the RW serializer sources it is not checked.

UPDATE:

Okay, the issue is a little bit different.

In our codebase we are calling `self.get_wirte_serializer` (note that we do not have `write_serializer_class` attribute, we do not have `get_write_serializer_class` method, we rely only on `get_serializer_class` in this case.

Stack trace is the following:
```python
  File "/home/user/app/src/application/view.py", line 123, in some_method
    serializer = self.get_write_serializer(data=data)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/venv/lib/python3.11/site-packages/drf_rw_serializers/generics.py", line 96, in get_write_serializer
    serializer_class = self.get_write_serializer_class()
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/venv/lib/python3.11/site-packages/drf_rw_serializers/generics.py", line 109, in get_write_serializer_class
    return self._get_serializer_class()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/venv/lib/python3.11/site-packages/drf_rw_serializers/generics.py", line 23, in _get_serializer_class
    self.serializer_class is not None
```

Please note that the error is about `read` somehow:
```
'MyApiViewSet' should either include one of `serializer_class` and `read_serializer_class` attribute, or override one of the `get_serializer_class()`, `get_read_serializer_class()` method.
```

### PR DESCRIPTION
**Pull request by pamella**
**Description:** Describe in a couple of sentence what this PR adds

Refactor strategy to handle `RecursionError` between `get_serializer_class`, `get_read_serializer`, and `get_write_serializer`

Relates to PR #77

**Dependencies:** dependencies on other outstanding PRs, issues, etc. 

Resolves #86 

**Merge deadline:** List merge deadline (if any)

**Installation instructions:** List any non-trivial installation 
instructions.

**Testing instructions:**

[Screencast from 05-06-2024 15:49:26.webm](https://github.com/vintasoftware/drf-rw-serializers/assets/22326737/3a9f768c-1e02-4aab-a787-df93f176218b)

[Screencast from 05-06-2024 15:52:11.webm](https://github.com/vintasoftware/drf-rw-serializers/assets/22326737/bc4d6058-2f88-4840-b680-971c4db9269d)

**Reviewers:**
- [ ] tag reviewer 
- [ ] tag reviewer 

**Merge checklist:**
- [ ] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [ ] Commits are squashed
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.

### CODE REVIEW

File: drf_rw_serializers/generics.py [hunk 1]
```diff
@@ -52,7 +36,8 @@ def get_serializer_class(self):
                     "attribute, or override the `get_read_serializer_class()` or "
                     "`get_serializer_class()` method." % self.__class__.__name__
                 )
-                return self.get_read_serializer_class()
+                # `use_serializer_class` is used to prevent a `RecursionError`
+                return self.get_read_serializer_class(use_serializer_class=True)
```
Comment (pamella) [inline drf_rw_serializers/generics.py]: Let me know if you think another name (`use_serializer_class`) is better.

File: test_settings.py [hunk 2]
```diff
@@ -38,3 +38,5 @@ def root(*args):
 ROOT_URLCONF = "example_app.urls"
 
 SECRET_KEY = "insecure-secret-key"
+
+DEFAULT_AUTO_FIELD = "django.db.models.AutoField"
```
Comment (pamella) [inline test_settings.py:42]: Drive-by warning fix.

File: tests/test_generics.py [hunk 3]
```diff
@@ -61,6 +61,21 @@ def test_serializer_class_not_provided(self):
...
+
+        self.assertEqual(
+            GetSerializerClassView().get_serializer_class(), OrderedMealDetailsSerializer
+        )
+        self.assertEqual(
+            GetSerializerClassView().get_read_serializer_class(), OrderedMealDetailsSerializer
+        )
+        self.assertEqual(
+            GetSerializerClassView().get_write_serializer_class(), OrderedMealDetailsSerializer
+        )
```
Comment (pamella) [inline tests/test_generics.py:77]: Regression test.

File: drf_rw_serializers/generics.py [hunk 4]
```diff
@@ -75,16 +70,21 @@ def get_read_serializer(self, *args, **kwargs):
...
         serializations depending on the incoming request.
+
         (Eg. admins get full serialization, others get basic serialization)
         """
         if getattr(self, "read_serializer_class", None) is None:
-            return self._get_serializer_class()
+            if use_serializer_class:
+                return self.serializer_class
+
+            return self.get_serializer_class()
```
Comment (pamella) [inline drf_rw_serializers/generics.py]: That's the approach I found to avoid raising a `RecursionError` while allowing:
1. `get_serializer_class` refer to `get_read_serializer_class` and `get_write_serializer_class` 
2. `get_read_serializer_class` and `get_write_serializer_class` refer back to `get_serializer_class`

File: drf_rw_serializers/generics.py (hunk 1 above)
Comment (fjsj) [reply drf_rw_serializers/generics.py]: use_serializer_class -> default_to_serializer_class perhaps?

**Review by fjsj — CHANGES_REQUESTED** [review]
Overall LGTM, please check comment.

File: drf_rw_serializers/generics.py (hunk 1 above)
Comment (pamella) [reply drf_rw_serializers/generics.py]: Done, thanks

**Review by fjsj — APPROVED** [review]

File: drf_rw_serializers/generics.py (hunk 4 above)
Comment (sshishov) [reply drf_rw_serializers/generics.py]: Hi @pamella @fjsj ,

Okay guys, apparently your `RecursionError` fix will break potentially everyone who override `get_read_serializer_class` or `get_write_serializer_class`.

```
TypeError: MyViewSet.get_read_serializer_class() got an unexpected keyword argument 'default_to_serializer_class'
```

When we override before, we never include any arguments like:
```python
@typing_extensions.override
def get_read_serializer_class(self) -> drf_serializers.BaseSerializer[MyModel]:
    if self.action == 'my_action`:
        return MySerializer
    else:
        return super().get_read_serializer_class()
```

With this change everything become broken... I do not like the fix, also I would like to understand why do we have this "infinite recursion"? In version `1.1.1` everything was working and this workaround was not needed. What has changed since then?

Comment (pamella) [reply drf_rw_serializers/generics.py]: > _In version 1.1.1 everything was working and this workaround was not needed. What has changed since then?_

@sshishov Please check the main PR that implemented the breaking changes, [Update get_serializer_class to consider the request method #77](https://github.com/vintasoftware/drf-rw-serializers/pull/77). Its description contains more information about the issues it was meant to address.

Comment (fjsj) [reply drf_rw_serializers/generics.py]: @sshishov you need to accept *args, **kwargs in your `def get_read_serializer_class`. We'll update the Usage docs to mention this. Unfortunately we see no other way of supporting the browsable API w/o the changes introduced in this PR #87.

```python
@typing_extensions.override
def get_read_serializer_class(self, *args, **kwargs) -> drf_serializers.BaseSerializer[MyModel]:
    if self.action == 'my_action`:
        return MySerializer
    else:
        return super().get_read_serializer_class(*args, **kwargs)
```
